File: btc-treasury/src/reporter.rs
use std::sync::Arc;

use tokio::time::{interval, Duration};

use crate::account_spec::ExchangeKind;
use crate::format::{escape_mdv2, send_mdv2_safe};
use crate::scanner::{ScannerState, RecentDecision};
use crate::memory::MemoryStore;

/// When there are more pairs than this threshold, the report switches from
/// per-pair lines to a compact summary-only mode to stay within Telegram's
/// 4096-char message limit. `send_mdv2_safe` will chunk even if this limit
/// is reached, but keeping individual messages meaningful is better UX.
const PAIR_DETAIL_THRESHOLD: usize = 20;

/// Maximum chars for a `reason` field in Recent Decisions before we truncate.
const MAX_REASON_CHARS: usize = 120;

/// Per-account, per-exchange report target. One reporter loop iterates
/// `Vec<PerAccountReport>` and emits a per-binding block. With one `default`
/// account on a single exchange the output is byte-identical to the
/// pre-Fase-1.5 reporter. With N accounts × M exchanges each binding gets
/// its own block prefixed `[id/exchange]`.
#[derive(Clone)]
pub struct PerAccountReport {
    pub account_id: String,
    pub exchange: ExchangeKind,
    pub state: Arc<ScannerState>,
    pub mem: Arc<MemoryStore>,
    /// Per-account chat IDs (from `AccountSpec.telegram_chat_ids`). If empty,
    /// the reporter falls back to `fallback_chat_ids` (the legacy global
    /// `telegram_report_chat_ids` env). This preserves single-account behavior
    /// where the global list is the only source.
    pub chat_ids: Vec<i64>,
}

pub async fn run(
    accounts: Vec<PerAccountReport>,
    bot_token: String,
    fallback_chat_ids: Vec<i64>,
    interval_mins: u64,
) {
    // Resolve effective chat targets per account (per-account → fallback).
    let effective: Vec<(PerAccountReport, Vec<i64>)> = accounts
        .into_iter()
        .map(|a| {
            let chats = if a.chat_ids.is_empty() {
                fallback_chat_ids.clone()
            } else {
                a.chat_ids.clone()
            };
            (a, chats)
        })
        .filter(|(_, chats)| !chats.is_empty())
        .collect();

    if effective.is_empty() {
        tracing::warn!("No report chat IDs configured — reporter disabled");
        return;
    }

    // Decide the title mode for each binding up front. Three modes:
    //   - `Legacy`: only one binding overall → "*BTC Scan Report — {exchange} Spot*"
    //     (preserves pre-Fase-1.5 byte-for-byte output)
    //   - `MultiAccount { id }`: multiple distinct ids, each single exchange
    //     → "*BTC Scan Report — [{id}]*"
    //   - `MultiExchange { id, exchange }`: same id has multiple exchanges
    //     → "*BTC Scan Report — [{id}/{exchange}]*"
    let total = effective.len();
    let distinct_ids: std::collections::HashSet<String> =
        effective.iter().map(|(a, _)| a.account_id.clone()).collect();
    let bindings_per_id = count_bindings_per_id(&effective);
    let multi = total > 1;

    tracing::info!(
        "BTC reporter started (every {} min) for {} binding(s) across {} account(s)",
        interval_mins, total, distinct_ids.len()
    );

    let mut tick = interval(Duration::from_secs(interval_mins * 60));
    // Lesson counters keyed by (account_id, exchange) so two bindings under
    // the same id don't share the same delta-detection window.
    let mut last_lesson_count: std::collections::HashMap<(String, ExchangeKind), usize> = effective
        .iter()
        .map(|(a, _)| ((a.account_id.clone(), a.exchange), a.mem.get_lessons().len()))
        .collect();

    loop {
        tick.tick().await;

        let bot = teloxide::prelude::Bot::new(&bot_token);
        for (acct, chat_ids) in &effective {
            let prev = last_lesson_count
                .get(&(acct.account_id.clone(), acct.exchange))
                .copied()
                .unwrap_or(0);

            // Async fetches (reporter is itself async).
            let snapshots = acct.state.all_snapshots().await;
            let recent: Vec<RecentDecision> = {
                let recents = acct.state.recent_decisions.read().await;
                recents.iter().rev().take(5).cloned().collect()
            };
            let all_lessons = acct.mem.get_lessons();
            let new_lessons: Vec<String> = if all_lessons.len() > prev {
                all_lessons[prev..].to_vec()
            } else {
                vec![]
            };

            // Skip empty reports to avoid Telegram spam.
            if snapshots.is_empty() && recent.is_empty() && new_lessons.is_empty() {
                continue;
            }

            // Per-binding title mode.
            let title = if total == 1 {
                ReportTitle::Legacy(acct.exchange.as_str())
            } else if bindings_per_id.get(&acct.account_id).copied().unwrap_or(1) > 1 {
                ReportTitle::MultiExchange(&acct.account_id, acct.exchange.as_str())
            } else {
                ReportTitle::MultiAccount(&acct.account_id)
            };

            let msg = format_report(title, &snapshots, &recent, &new_lessons);
            for chat_id in chat_ids {
                let chat = teloxide::prelude::ChatId(*chat_id);
                if let Err(e) = send_mdv2_safe(&bot, chat, &msg).await {
                    tracing::error!(
                        "Reporter [{}/{}]: failed to send to {}: {}",
                        acct.account_id, acct.exchange.as_str(), chat_id, e
                    );
                }
            }
        }
        // Update lesson counters AFTER the per-account send so the next tick
        // detects only genuinely new lessons.
        for (acct, _) in &effective {
            last_lesson_count.insert(
                (acct.account_id.clone(), acct.exchange),
                acct.mem.get_lessons().len(),
            );
        }

        // Multi-binding aggregate footer (sent once per loop to the first
        // binding's chat list to avoid spamming every chat with the same
        // digest).
        if multi {
            if let Some((_, first_chats)) = effective.first() {
                let total_btc: f64 = effective
                    .iter()
                    .map(|(a, _)| a.mem.get_treasury_state().current_btc)
                    .sum();
                let total_vault: f64 = effective
                    .iter()
                    .map(|(a, _)| a.mem.get_treasury_state().btc_treasury_vault)
                    .sum();
                let total_trades: u64 = effective
                    .iter()
                    .map(|(a, _)| a.mem.get_treasury_state().total_trades as u64)
                    .sum();
                let aggregate = format!(
                    "\n──────────\n*Aggregate — All Bindings*\nBTC: {:.8} \\| Vault: {:.8} \\| Trades: {}",
                    total_btc, total_vault, total_trades
                );
                for chat_id in first_chats {
                    let chat = teloxide::prelude::ChatId(*chat_id);
                    if let Err(e) = send_mdv2_safe(&bot, chat, &aggregate).await {
                        tracing::error!("Reporter [aggregate]: failed to send to {}: {}", chat_id, e);
                    }
                }
            }
        }
    }
}

/// How a report block should title itself. `Legacy` is the pre-Fase-1.5
/// single-binding title (no `[id]` prefix). `MultiAccount` and
/// `MultiExchange` add prefixes so multi-binding users can tell blocks apart.
#[derive(Debug, Clone, Copy)]
enum ReportTitle<'a> {
    Legacy(&'a str),                    // "*BTC Scan Report — {exchange} Spot*"
    MultiAccount(&'a str),              // "*BTC Scan Report — [{id}]*"
    MultiExchange(&'a str, &'a str),    // "*BTC Scan Report — [{id}/{exchange}]*"
}

fn count_bindings_per_id(
    effective: &[(PerAccountReport, Vec<i64>)],
) -> std::collections::HashMap<String, usize> {
    let mut map: std::collections::HashMap<String, usize> = std::collections::HashMap::new();
    for (a, _) in effective {
        *map.entry(a.account_id.clone()).or_insert(0) += 1;
    }
    map
}

fn format_report(
    title: ReportTitle<'_>,
    snapshots: &[crate::scanner::PairSnapshot],
    recent: &[RecentDecision],
    new_lessons: &[String],
) -> String {
    let mut lines: Vec<String> = Vec::new();

    // Single-binding legacy: "*BTC Scan Report — {exchange} Spot*" (e.g.
    // "Binance Spot"). This is byte-identical to pre-Fase-1.5 for the
    // single-account user.
    match title {
        ReportTitle::Legacy(exchange) => {
            lines.push(format!("*BTC Scan Report — {} Spot*\n", exchange));
        }
        ReportTitle::MultiAccount(id) => {
            lines.push(format!("*BTC Scan Report — [{}]*\n", id));
        }
        ReportTitle::MultiExchange(id, exchange) => {
            lines.push(format!("*BTC Scan Report — [{}/{}]*\n", id, exchange));
        }
    }

    let total_scanned: u64 = snapshots.iter().map(|s| s.stats.scanned).sum();
    let total_errors: u64 = snapshots.iter().map(|s| s.stats.errors).sum();

    if total_scanned == 0 && recent.is_empty() && new_lessons.is_empty() {
        lines.push("No scans in this period\\.".into());
        return lines.join("\n");
    }

    if !snapshots.is_empty() {
        // Count per-recommendation bucket for compact summary
        let approve_cnt = snapshots.iter().filter(|s| s.last_recommendation == "APPROVE").count();
        let monitor_cnt = snapshots.iter().filter(|s| s.last_recommendation == "MONITOR").count();
        let protect_cnt = snapshots.iter().filter(|s| {
            matches!(s.last_recommendation.as_str(), "PROTECT_TREASURY" | "ENABLE_SAFE_MODE" | "REDUCE_EXPOSURE")
        }).count();
        let reject_cnt = snapshots.iter().filter(|s| {
            !s.last_recommendation.is_empty()
                && !matches!(s.last_recommendation.as_str(),
                    "APPROVE" | "MONITOR" | "PROTECT_TREASURY" | "ENABLE_SAFE_MODE" | "REDUCE_EXPOSURE")
        }).count();

        lines.push(format!(
            "Pairs: {} \\| Scans: {} \\| Errors: {}\n✅ {} \\| 👁 {} \\| 🛡 {} \\| ❌ {}\n",
            snapshots.len(), total_scanned, total_errors,
            approve_cnt, monitor_cnt, protect_cnt, reject_cnt
        ));

        if snapshots.len() <= PAIR_DETAIL_THRESHOLD {
            // ── Detailed mode: one line per pair ────────────────────────
            for s in snapshots {
                let icon = match s.last_recommendation.as_str() {
                    "APPROVE" => "✅",
                    "MONITOR" => "👁",
                    "PROTECT_TREASURY" | "ENABLE_SAFE_MODE" | "REDUCE_EXPOSURE" => "🛡",
                    _ if s.last_recommendation.is_empty() => "⏳",
                    _ => "❌",
                };
                lines.push(format!(
                    "{} {} {} scans \\| {} \\(conf: {:.2}\\)",
                    icon,
                    escape_mdv2(&s.pair),
                    s.stats.scanned,
                    escape_mdv2(&s.last_recommendation),
                    s.last_confidence,
                ));
            }
        } else {
            // ── Compact mode: only show top APPROVE and notable pairs ───
            // Show up to 5 APPROVE pairs (highest confidence first)
            let mut approve_pairs: Vec<&crate::scanner::PairSnapshot> = snapshots
                .iter()
                .filter(|s| s.last_recommendation == "APPROVE")
                .collect();
            approve_pairs.sort_by(|a, b| b.last_confidence.partial_cmp(&a.last_confidence).unwrap_or(std::cmp::Ordering::Equal));

            if !approve_pairs.is_empty() {
                lines.push("*Top APPROVE pairs:*".into());
                for s in approve_pairs.iter().take(5) {
                    lines.push(format!(
                        "✅ {} conf:{:.2}",
                        escape_mdv2(&s.pair),
                        s.last_confidence,
                    ));
                }
            }
        }
    }

    if !recent.is_empty() {
        lines.push("\n*Recent Decisions:*".into());
        for (i, d) in recent.iter().enumerate() {
            let status_icon = match d.recommendation.as_str() {
                "APPROVE" => "✅",
                "MONITOR" => "👁",
                "PROTECT_TREASURY" | "ENABLE_SAFE_MODE" | "REDUCE_EXPOSURE" => "🛡",
                _ => "❌",
            };
            let time_short = if d.timestamp.len() > 16 {
                &d.timestamp[11..19]
            } else {
                &d.timestamp
            };
            // Truncate reason to avoid huge messages
            let reason_short = if d.reason.len() > MAX_REASON_CHARS {
                format!("{}…", &d.reason[..MAX_REASON_CHARS])
            } else {
                d.reason.clone()
            };
            lines.push(format!(
                "{}\\. {} {} {} \\- {} \\(conf: {:.2}, risk: {}\\)\n  \\_`{}`",
                i + 1,
                escape_mdv2(time_short),
                escape_mdv2(&d.pair),
                status_icon,
                escape_mdv2(&d.recommendation),
                d.confidence,
                escape_mdv2(&d.risk_level),
                escape_mdv2(&reason_short),
            ));
        }
    }

    if !new_lessons.is_empty() {
        lines.push("\n*New Lessons:*".into());
        for (i, lesson) in new_lessons.iter().take(3).enumerate() {
            let short = if lesson.len() > 150 {
                format!("{}...", &lesson[..147])
            } else {
                lesson.clone()
            };
            lines.push(format!(
                "{}{}\\. {}",
                i + 1,
                r"\.",
                escape_mdv2(&short),
            ));
        }
    }

    lines.join("\n")
}

File: btc-treasury/internal/reporter/reporter.go
package reporter

import (
	"context"
	"fmt"
	"log"
	"sort"
	"strings"
	"time"

	"btc-treasury/internal/config"
	"btc-treasury/internal/memory"
	"btc-treasury/internal/scanner"
	"btc-treasury/internal/utils"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
)

const (
	pairDetailThreshold = 20
	maxReasonChars      = 120
)

type PerAccountReport struct {
	AccountID string
	Exchange  config.ExchangeKind
	State     *scanner.ScannerState
	Mem       memory.Store
	ChatIDs   []int64
}

type reportTitle struct {
	Legacy        bool
	MultiAccount  bool
	MultiExchange bool
	AccountID     string
	Exchange      string
}

func (t reportTitle) Format() string {
	if t.Legacy {
		return fmt.Sprintf("*BTC Scan Report — %s Spot*\n", t.Exchange)
	}
	if t.MultiExchange {
		return fmt.Sprintf("*BTC Scan Report — [%s/%s]*\n", t.AccountID, t.Exchange)
	}
	return fmt.Sprintf("*BTC Scan Report — [%s]*\n", t.AccountID)
}

func Run(
	ctx context.Context,
	accounts []PerAccountReport,
	botToken string,
	fallbackChatIDs []int64,
	intervalMins uint64,
) {
	var effective []struct {
		Report  PerAccountReport
		ChatIDs []int64
	}

	for _, a := range accounts {
		chats := a.ChatIDs
		if len(chats) == 0 {
			chats = fallbackChatIDs
		}
		if len(chats) > 0 {
			effective = append(effective, struct {
				Report  PerAccountReport
				ChatIDs []int64
			}{
				Report:  a,
				ChatIDs: chats,
			})
		}
	}

	if len(effective) == 0 {
		log.Printf("Reporter: No report chat IDs configured — reporter disabled")
		return
	}

	total := len(effective)
	distinctIDs := make(map[string]bool)
	for _, entry := range effective {
		distinctIDs[entry.Report.AccountID] = true
	}

	bindingsPerID := countBindingsPerID(effective)
	multi := total > 1

	log.Printf("BTC reporter started (every %d min) for %d binding(s) across %d account(s)",
		intervalMins, total, len(distinctIDs))

	bot, err := tgbotapi.NewBotAPI(botToken)
	if err != nil {
		log.Printf("Reporter: Failed to initialize Telegram bot: %v", err)
		return
	}

	lastLessonCount := make(map[string]int)
	for _, entry := range effective {
		key := fmt.Sprintf("%s|%s", entry.Report.AccountID, entry.Report.Exchange)
		lastLessonCount[key] = len(entry.Report.Mem.GetLessons())
	}

	ticker := time.NewTicker(time.Duration(intervalMins) * time.Minute)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			log.Printf("Reporter: Stopping loop")
			return
		case <-ticker.C:
			for _, entry := range effective {
				acct := entry.Report
				key := fmt.Sprintf("%s|%s", acct.AccountID, acct.Exchange)
				prev := lastLessonCount[key]

				snapshots := acct.State.AllSnapshots()
				
				recentDecisions := acct.State.GetRecentDecisions()
				var recent []scanner.RecentDecision
				for i := len(recentDecisions) - 1; i >= 0 && len(recent) < 5; i-- {
					recent = append(recent, recentDecisions[i])
				}

				allLessons := acct.Mem.GetLessons()
				var newLessons []string
				if len(allLessons) > prev {
					newLessons = allLessons[prev:]
				}

				if len(snapshots) == 0 && len(recent) == 0 && len(newLessons) == 0 {
					continue
				}

				var title reportTitle
				if total == 1 {
					title = reportTitle{Legacy: true, Exchange: strings.Title(string(acct.Exchange))}
				} else if bindingsPerID[acct.AccountID] > 1 {
					title = reportTitle{MultiExchange: true, AccountID: acct.AccountID, Exchange: string(acct.Exchange)}
				} else {
					title = reportTitle{MultiAccount: true, AccountID: acct.AccountID}
				}

				msg := formatReport(title, snapshots, recent, newLessons)

				for _, chatID := range entry.ChatIDs {
					_, err := utils.SendMdv2Safe(bot, chatID, msg)
					if err != nil {
						log.Printf("Reporter [%s/%s]: failed to send to %d: %v",
							acct.AccountID, acct.Exchange, chatID, err)
					}
				}
			}

			for _, entry := range effective {
				acct := entry.Report
				key := fmt.Sprintf("%s|%s", acct.AccountID, acct.Exchange)
				lastLessonCount[key] = len(acct.Mem.GetLessons())
			}

			if multi {
				if len(effective) > 0 {
					firstEntry := effective[0]
					var totalBTC float64
					var totalVault float64
					var totalTrades int

					for _, entry := range effective {
						state := entry.Report.Mem.GetTreasuryState()
						totalBTC += state.CurrentBtc
						totalVault += state.BtcTreasuryVault
						totalTrades += state.TotalTrades
					}

					aggregateMsg := fmt.Sprintf(
						"\n──────────\n*Aggregate — All Bindings*\nBTC: %.8f \\| Vault: %.8f \\| Trades: %d",
						totalBTC, totalVault, totalTrades,
					)

					for _, chatID := range firstEntry.ChatIDs {
						_, err := utils.SendMdv2Safe(bot, chatID, aggregateMsg)
						if err != nil {
							log.Printf("Reporter [aggregate]: failed to send to %d: %v", chatID, err)
						}
					}
				}
			}
		}
	}
}

func countBindingsPerID(effective []struct {
	Report  PerAccountReport
	ChatIDs []int64
}) map[string]int {
	m := make(map[string]int)
	for _, entry := range effective {
		m[entry.Report.AccountID]++
	}
	return m
}

func formatReport(
	title reportTitle,
	snapshots []scanner.PairSnapshot,
	recent []scanner.RecentDecision,
	newLessons []string,
) string {
	var lines []string

	lines = append(lines, title.Format())

	var totalScanned uint64
	var totalErrors uint64
	for _, s := range snapshots {
		totalScanned += s.Stats.Scanned
		totalErrors += s.Stats.Errors
	}

	if totalScanned == 0 && len(recent) == 0 && len(newLessons) == 0 {
		lines = append(lines, "No scans in this period\\.")
		return strings.Join(lines, "\n")
	}

	if len(snapshots) > 0 {
		var approveCnt, monitorCnt, protectCnt, rejectCnt int
		for _, s := range snapshots {
			switch s.LastRecommendation {
			case "APPROVE":
				approveCnt++
			case "MONITOR":
				monitorCnt++
			case "PROTECT_TREASURY", "ENABLE_SAFE_MODE", "REDUCE_EXPOSURE":
				protectCnt++
			default:
				if s.LastRecommendation != "" {
					rejectCnt++
				}
			}
		}

		lines = append(lines, fmt.Sprintf(
			"Pairs: %d \\| Scans: %d \\| Errors: %d\n✅ %d \\| 👁 %d \\| 🛡 %d \\| ❌ %d\n",
			len(snapshots), totalScanned, totalErrors,
			approveCnt, monitorCnt, protectCnt, rejectCnt,
		))

		if len(snapshots) <= pairDetailThreshold {
			for _, s := range snapshots {
				var icon string
				switch s.LastRecommendation {
				case "APPROVE":
					icon = "✅"
				case "MONITOR":
					icon = "👁"
				case "PROTECT_TREASURY", "ENABLE_SAFE_MODE", "REDUCE_EXPOSURE":
					icon = "🛡"
				default:
					if s.LastRecommendation == "" {
						icon = "⏳"
					} else {
						icon = "❌"
					}
				}

				lines = append(lines, fmt.Sprintf(
					"%s %s %d scans \\| %s \\(conf: %.2f\\)",
					icon,
					utils.EscapeMdv2(s.Pair),
					s.Stats.Scanned,
					utils.EscapeMdv2(s.LastRecommendation),
					s.LastConfidence,
				))
			}
		} else {
			var approvePairs []scanner.PairSnapshot
			for _, s := range snapshots {
				if s.LastRecommendation == "APPROVE" {
					approvePairs = append(approvePairs, s)
				}
			}

			sort.Slice(approvePairs, func(i, j int) bool {
				return approvePairs[i].LastConfidence > approvePairs[j].LastConfidence
			})

			if len(approvePairs) > 0 {
				lines = append(lines, "*Top APPROVE pairs:*")
				limit := 5
				if len(approvePairs) < 5 {
					limit = len(approvePairs)
				}
				for i := 0; i < limit; i++ {
					s := approvePairs[i]
					lines = append(lines, fmt.Sprintf(
						"✅ %s conf:%.2f",
						utils.EscapeMdv2(s.Pair),
						s.LastConfidence,
					))
				}
			}
		}
	}

	if len(recent) > 0 {
		lines = append(lines, "\n*Recent Decisions:*")
		for i, d := range recent {
			var statusIcon string
			switch d.Recommendation {
			case "APPROVE":
				statusIcon = "✅"
			case "MONITOR":
				statusIcon = "👁"
			case "PROTECT_TREASURY", "ENABLE_SAFE_MODE", "REDUCE_EXPOSURE":
				statusIcon = "🛡"
			default:
				statusIcon = "❌"
			}

			timeShort := d.Timestamp
			if len(d.Timestamp) > 16 {
				timeShort = d.Timestamp[11:19]
			}

			reasonShort := d.Reason
			if len(d.Reason) > maxReasonChars {
				reasonShort = d.Reason[:maxReasonChars] + "…"
			}

			lines = append(lines, fmt.Sprintf(
				"%d\\. %s %s %s \\- %s \\(conf: %.2f, risk: %s\\)\n  \\_`%s`",
				i+1,
				utils.EscapeMdv2(timeShort),
				utils.EscapeMdv2(d.Pair),
				statusIcon,
				utils.EscapeMdv2(d.Recommendation),
				d.Confidence,
				utils.EscapeMdv2(d.RiskLevel),
				utils.EscapeMdv2(reasonShort),
			))
		}
	}

	if len(newLessons) > 0 {
		lines = append(lines, "\n*New Lessons:*")
		limit := 3
		if len(newLessons) < 3 {
			limit = len(newLessons)
		}
		for i := 0; i < limit; i++ {
			lesson := newLessons[i]
			short := lesson
			if len(lesson) > 150 {
				short = lesson[:147] + "..."
			}
			lines = append(lines, fmt.Sprintf(
				"%d\\. %s",
				i+1,
				utils.EscapeMdv2(short),
			))
		}
	}

	return strings.Join(lines, "\n")
}
